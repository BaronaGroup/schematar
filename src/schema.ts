import ObjectId from './object-id'
import Complex from './complex'

export type DefaultContext = 'mongoose' | 'typescript' | 'jsonschema'

export interface Schema {
    fields: SchemaFields
}

type SimpleType = Symbol | Complex | typeof String | typeof Number | typeof Boolean | typeof Date | typeof Object
type IndexOptions = false | true | 'unique' | 'unique-sparse'

export type PlainType = SimpleType | [FieldInfo | SimpleType]

export type Field = FieldInfo | PlainType

export interface FieldInfo {
    type: PlainType
    index?: IndexOptions
    format?: any
    optional?: boolean
    presentIn?: Array<string>
    optionalIn?: Array<string>
    enum?: string[]
    mongooseDefault?: any
    mongooseRef?: string
}

export interface SchemaFields {
    [name: string]: Field
}
