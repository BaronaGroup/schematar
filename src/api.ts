import fs from 'fs'
import path from 'path'

import mkdirp from 'mkdirp'

import ComplexType from './complex'
import { generateHash } from './hash-schema'
import jsonSchemaOutput, { JSONSchemaOptions } from './json-schema-output'
import { karhu } from './karhu'
import mongooseOutput from './mongoose-output'
import nowToken from './now'
import { ObjectId as ObjectIdType } from './object-id'
import { Schema as SchemaType } from './schema'
import tsOutput, { TSOptions } from './ts-output'

export const ObjectId = ObjectIdType
export type Schema = SchemaType
export const Complex = ComplexType
export const now = nowToken
const log = karhu('schematar-api')

export function createMongooseSchema(schema: Schema, context = 'mongoose') {
  return mongooseOutput(schema, context)
}

export function createTypescriptInterfaceDefinition(
  exportName: string,
  schema: Schema | ComplexType,
  context: string = 'typescript',
  options: TSOptions = {}
) {
  return tsOutput(exportName, schema, context, options)
}

export function createJSONSchema(schema: Schema, context: string = 'jsonschema', options: JSONSchemaOptions = {}) {
  return jsonSchemaOutput(schema, context, options)
}

export async function createTypescriptInterfaceFiles(
  sourceFileGlobOrFileArray: string | string[],
  outputPath: string,
  logCreations = false
) {
  const { default: glob } = await import('glob')
  await new Promise((resolve, reject) => mkdirp(outputPath, (err) => (!err ? resolve() : reject(err))))
  const files = [].concat(
    ...(Array.isArray(sourceFileGlobOrFileArray)
      ? sourceFileGlobOrFileArray.map((f) => glob.sync(f))
      : [glob.sync(sourceFileGlobOrFileArray)])
  )
  for (const file of files) {
    await createTSInterfaceFile(file, outputPath, logCreations)
  }
}

export interface TypescriptSchemaDefinition {
  name: string
  context?: string
  omitExtraExports?: boolean
  schema?: Schema | ComplexType
  exportHash?: string
  doNotImportObjectId?: boolean
}

interface SchemaFile {
  name?: string
  default?: Schema
  schema?: Schema
  typescriptSchemas?: TypescriptSchemaDefinition[]
}

export async function createTSInterfaceFile(
  filename: string,
  outputPath: string,
  logCreations: boolean = true,
  formatOutputFilename = defaultOutputFilenameFormatter
) {
  const schemaFile = require(path.resolve(filename)) as SchemaFile
  const topLevelSchema = schemaFile.schema || schemaFile.default
  const topLevelName = schemaFile.name || pickNameFromFilename(filename)
  const exported: string[] = []
  if (schemaFile.typescriptSchemas) {
    const typescriptSchemas = schemaFile.typescriptSchemas
    for (const schemaDef of typescriptSchemas) {
      const schema = schemaDef.schema || topLevelSchema,
        name = schemaDef.name || topLevelName

      if (!schema) throw new Error('No schema found for ' + name)
      exported.push(
        createTypescriptInterfaceDefinition(name, schema, schemaDef.context || 'typescript', {
          ...schemaDef,
          doNotImportObjectId: schemaDef.doNotImportObjectId || typescriptSchemas.indexOf(schemaDef) > 0,
        })
      )
    }
  } else if (topLevelSchema) {
    exported.push(createTypescriptInterfaceDefinition(topLevelName, topLevelSchema))
  }
  const outputFilename = path.join(outputPath, formatOutputFilename(filename))

  if (logCreations) {
    log.info('Storing typescript interfaces in', outputFilename)
  }

  if (exported.length) {
    await new Promise((resolve, reject) =>
      fs.writeFile(outputFilename, exported.join('\n'), 'utf8', (err) => (!err ? resolve() : reject(err)))
    )
  }
}

function defaultOutputFilenameFormatter(inputFilename: string) {
  const ext = path.extname(inputFilename),
    basename = path.basename(inputFilename, ext)

  return basename + '.schematar.ts'
}

function pickNameFromFilename(filename: string) {
  const fn = path.basename(filename, path.extname(filename))
  return (fn[0].toUpperCase() + fn.substring(1))
    .replace(/[\-_][a-z]/g, (match) => match[1].toUpperCase())
    .replace(/[^a-zA-Z0-9]/g, '')
}

export function hashSchema(schema: Schema | ComplexType) {
  return generateHash(schema)
}
