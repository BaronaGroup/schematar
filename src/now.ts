export default 'now-b9759814-a87f-4193-bbf8-c1003db19515'
