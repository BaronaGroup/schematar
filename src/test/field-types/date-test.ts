import { createMongooseSchema } from '../../api'
import now from '../../now'
import { Schema } from '../../schema'
import { testJSONSchema, testMongooseField, testTypescriptInterface } from '../test-utils'

describe('date-test', function () {
  describe('basic type', function () {
    const schema: Schema = {
      fields: {
        field: Date,
      },
    }

    it(
      'json-schema',
      testJSONSchema(schema, (data) => {
        expect(data).toMatchSnapshot()
      })
    )

    it(
      'mongoose',
      testMongooseField(schema, 'field', (field) => {
        expect(field).toEqual({ type: Date })
      })
    )

    it(
      'typescript',
      testTypescriptInterface(schema, (tsInterface) => {
        expect(tsInterface).toMatchSnapshot()
      })
    )
  })

  it('defaulting to "now" in mongoose', function () {
    const schema = {
      fields: {
        field: { type: Date, mongooseDefault: now },
      },
    }
    const ms = createMongooseSchema(schema)
    expect(ms.field.default).toBe(Date.now)
  })
})
