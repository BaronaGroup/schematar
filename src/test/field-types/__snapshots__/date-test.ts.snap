// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`date-test basic type json-schema 1`] = `
"{
  \\"type\\": \\"object\\",
  \\"$schema\\": \\"http://json-schema.org/draft-07/schema#\\",
  \\"properties\\": {
    \\"field\\": {
      \\"type\\": \\"string\\",
      \\"pattern\\": \\"^\\\\\\\\d\\\\\\\\d\\\\\\\\d\\\\\\\\d-\\\\\\\\d\\\\\\\\d-\\\\\\\\d\\\\\\\\dT\\\\\\\\d\\\\\\\\d:\\\\\\\\d\\\\\\\\d:\\\\\\\\d\\\\\\\\d\\\\\\\\.\\\\\\\\d\\\\\\\\d\\\\\\\\d(Z|[+-]\\\\\\\\d\\\\\\\\d:?\\\\\\\\d\\\\\\\\d)$\\"
    }
  },
  \\"required\\": [
    \\"field\\"
  ]
}"
`;

exports[`date-test basic type typescript 1`] = `
"// Generated file, do not edit!

// tslint:disable array-type
// @ts-ignore -- ignore possibly unused type parameters
export interface TestInterfaceBase<IDType, DateType> {
  field: DateType
}"
`;
